src/wallet/wallet.cpp

```diff
@@ -4997,125 +5086,121 @@ void CWallet::GetFilteredNotes(
-                continue;
-            }
+        std::set<libzcash::SaplingPaymentAddress> SaplingAddrs;
+        GetSaplingPaymentAddresses(SaplingAddrs);
 
-            // skip note which has been spent
-            if (ignoreSpent && nd.nullifier && IsSproutSpent(*nd.nullifier)) {
-                continue;
-            }
+        if (sproutAddrs.size() > 0)
+            filterAddresses.insert(sproutAddrs.begin(), sproutAddrs.end());
+        if (SaplingAddrs.size() > 0)
+            filterAddresses.insert(SaplingAddrs.begin(), SaplingAddrs.end());
 
-            // skip notes which cannot be spent
-            if (requireSpendingKey && !HaveSproutSpendingKey(pa)) {
-                continue;
-            }
+        if (filterAddresses.size() < 1)
+            return;
+    }
 
-            // skip locked notes
-            if (ignoreLocked && IsLockedNote(jsop)) {
-                continue;
-            }
+    auto& index = instance->index.get<by_timestamp>();
 
-            int i = jsop.js; // Index into CTransaction.vJoinSplit
-            int j = jsop.n; // Index into JSDescription.ciphertexts
+    for (auto& zaddr : filterAddresses) {
+        uint256 paymentaddress_hash;
+        if (boost::get<libzcash::SproutPaymentAddress>(&zaddr) != nullptr) {
+            SproutPaymentAddress paymentaddress = *boost::get<libzcash::SproutPaymentAddress>(&zaddr);
+            paymentaddress_hash = paymentaddress.GetHash();
 
-            // Get cached decryptor
-            ZCNoteDecryption decryptor;
-            if (!GetNoteDecryptor(pa, decryptor)) {
-                // Note decryptors are created when the wallet is loaded, so it should always exist
-                throw std::runtime_error(strprintf("Could not find note decryptor for payment address %s", keyIO.EncodePaymentAddress(pa)));
-            }
+            auto itr_sprout = GetNotesByType(paymentaddress_hash, NoteType::sprout, timestamp);
+            auto itr = itr_sprout.begin();
+
+            while (itr != itr_sprout.end()) {
+                auto nd = mapWallet[itr->hash].mapSproutNoteData[*itr->jsop];
+                auto wtx = mapWallet[itr->hash];
+
+                JSOutPoint jsop = *itr->jsop;
+                auto ts = itr->timestamp;
 
-            // determine amount of funds in the note
-            auto hSig = wtx.vJoinSplit[i].h_sig(wtx.joinSplitPubKey);
-            try {
-                SproutNotePlaintext plaintext = SproutNotePlaintext::decrypt(
-                        decryptor,
-                        wtx.vJoinSplit[i].ciphertexts[j],
-                        wtx.vJoinSplit[i].ephemeralKey,
-                        hSig,
-                        (unsigned char) j);
+                ++itr;
+
+                filters.wtx = &wtx;
+                filters.nd_sprout = nd;
+                filters.jsop = jsop;
+                filters.pa_sprout = paymentaddress;
+                filters.cwallet = this;
+
+                if (filters.Common() || filters.Sprout())
```

Comment:
Obviously this needs to be changed if the sense of filter return values is changed, or with my suggested refactoring to pass the note information as arguments to the filter methods.

Reply:
https://github.com/zcash/zcash/pull/4399/commits/1431921924f32190b4f34f7c8cf1ea81d2957549